src/query.rs

```diff
@@ -8,113 +8,182 @@ use http::{
+            match response.data.get_result_url {
+                Some(result_url) => {
+                    response =
+                        poll_for_async_results(http, account, &result_url, session_token, timeout)
+                            .await?
+                }
+                None => {
+                    return Err(Error::NoPollingUrlAsyncQuery);
+                }
             }
-            None => {
-                return Err(Error::NoPollingUrlAsyncQuery);
+        }
+
+        if let Some(SESSION_EXPIRED) = response.code.as_deref() {
+            return Err(Error::SessionExpired);
+        }
+
+        if !response.success {
+            return Err(Error::Communication(response.message.unwrap_or_default()));
+        }
+
+        if let Some(format) = response.data.query_result_format {
+            if format != "json" {
+                return Err(Error::UnsupportedFormat(format.clone()));
             }
         }
-    }
 
-    if let Some(SESSION_EXPIRED) = response.code.as_deref() {
-        return Err(Error::SessionExpired);
+        let http = http.clone();
+        let qrmk = response.data.qrmk.unwrap_or_default();
+        let chunks = response.data.chunks.unwrap_or_default();
+        let row_types = response.data.row_types.ok_or_else(|| {
+            Error::UnsupportedFormat("the response doesn't contain 'rowtype'".to_string())
+        })?;
+        let row_set = response.data.row_set.ok_or_else(|| {
+            Error::UnsupportedFormat("the response doesn't contain 'rowset'".to_string())
+        })?;
+
+        let column_types = row_types
+            .into_iter()
+            .enumerate()
+            .map(|(i, row_type)| {
+                (
+                    row_type.name.to_ascii_uppercase(),
+                    SnowflakeColumnType {
+                        index: i,
+                        snowflake_type: row_type.data_type,
+                        nullable: row_type.nullable,
+                    },
+                )
+            })
+            .collect::<HashMap<_, _>>();
+        let column_types = Arc::new(column_types);
+
+        let chunk_headers = response.data.chunk_headers.unwrap_or_default();
+        let chunk_headers: HeaderMap = HeaderMap::try_from(&chunk_headers)?;
+
+        Ok(Self {
+            http,
+            qrmk,
+            chunks,
+            chunk_headers,
+            column_types,
+            row_set: Some(row_set),
+        })
     }
 
-    if !response.success {
-        return Err(Error::Communication(response.message.unwrap_or_default()));
+    /// Check if there are no more rows to fetch
+    pub fn eof(&self) -> bool {
+        self.row_set.is_none() && self.chunks.is_empty()
     }
 
-    if let Some(format) = response.data.query_result_format {
-        if format != "json" {
-            return Err(Error::UnsupportedFormat(format.clone()));
+    /// Fetch a single chunk
+    pub async fn fetch_chunk(&mut self) -> Result<Vec<SnowflakeRow>> {
```

Comment:
We might want to separately handle connection errors and successfully finished jobs. Could we update this function to be more like `Iterator::next`?

```rust
pub async fn fetch_next_chunk(&mut self) -> Result<Option<Vec<SnowflakeRow>>> {...}
```

This function returns:
- `Ok(None)` if there's no remaining chunk.
- `Ok(Some(rows))` if it successfully fetched the next chunk.
- `Err(e)` if it failed to fetch the next chunk.